### ISSUE TITLE
Modifying @Query has problems changing entities by their UUID

### ISSUE THREAD

**Issue by burdoto** [spring-projects/spring-data-jpa#2720]
Hello,

I ran into an issue regarding UUIDs in a JPA CrudRepository in my project;
Basically when writing an object with a UUID, the UUID is converted alright, but when trying to access the row by this UUID, that fails.
Unfortunately I don't have good example code just right now other than [a blob in the repository](https://github.com/comroid-git/comroid.org/tree/ab74814b4e71aeecff09a679363c66fa1bd0601f/auth/src/main) right now, but I will try to make a reproducing code example right away tomorrow.

About the repository so far:
Basically I am storing `UserAccount` objects by UUID in the `AccountRepository` table.
But when [trying to modify the account by UUID](https://github.com/comroid-git/comroid.org/blob/ab74814b4e71aeecff09a679363c66fa1bd0601f/auth/src/main/java/org/comroid/auth/repo/AccountRepository.java#L25), that never succeeded because the UUID was not converted correctly for the query, so I never got a result row to modify.

**Comment by spring-projects-issues** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by burdoto** [comment]
Sorry for the late response, I kinda forgot.

I've now created a little test project.
Consider the following Entity:
```java
```

**Comment by burdoto** [comment]
Sorry for the late response, I kinda forgot.

I've now created a little test project.
Consider the following Entity:
```java
package com.example.springjpauuidconverter;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.util.UUID;


@Entity
@Table(name = "uuid_conv")
public class TestEntity {
    @Id
    public UUID id;
    @Column
    public UUID id2;
    @Column
    public String name;

    public UUID getId() {
        return id;
    }

    public UUID getId2() {
        return id2;
    }

    public String getName() {
        return name;
    }

    public TestEntity() {
    }

    public TestEntity(String name) {
        this.id = UUID.randomUUID();
        this.id2 = UUID.randomUUID();
        this.name = name;
    }
}
```
Also, consider the following repository interface:
```java
package com.example.springjpauuidconverter;

import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;

import java.util.Optional;
import java.util.UUID;

public interface TestRepository extends CrudRepository<TestEntity, UUID> {
    @Query("select e from TestEntity e where e.id2 = ?1")
    Optional<TestEntity> findByAltId(UUID id);
}
```
Now, I've also prepared a little testing controller:
```java
package com.example.springjpauuidconverter;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;

import java.util.Optional;
import java.util.UUID;

@Controller
public class TestController {
    @Autowired
    public TestRepository repo;

    @GetMapping("/createEntities")
    public String initEntities() {
        repo.save(new TestEntity("apple"));
        repo.save(new TestEntity("beanz"));
        return "redirect:https://google.com";
    }

    @GetMapping("/get1/{id}")
    public Optional<TestEntity> getById1(@PathVariable("id") UUID id) {
        return repo.findById(id);
    }

    @GetMapping("/get2/{id}")
    public Optional<TestEntity> getById2(@PathVariable("id") UUID id) {
        return repo.findByAltId(id);
    }
}
```

That's all code I'm using (besides the main class, obviously).

Now, when launching this app and calling `localhost:8080/createEntities`, the entities are correctly created as defined in the endpoint:
![datagrip64_9lXFA5jMmR](https://user-images.githubusercontent.com/42473485/207868014-5d73b9d5-c50f-4d0d-a24f-6ec44bea5cd3.png)
Now, for accessing the `/get1/{id}` and `/get2/{id}` endpoints, I have set a debugging breakpoint to verify the UUID path variables are correctly parsed, which they are:
![idea64_loQWG3KzPn](https://user-images.githubusercontent.com/42473485/207868197-1b2fd7e1-912a-41b2-af9b-9e282a4198c4.png)
![idea64_gy5r66QOGA](https://user-images.githubusercontent.com/42473485/207868204-82fd62dc-028f-4b5a-a24b-6b65e4c6cf86.png)

However, when the code calls `repo.findById(id)`, the resulting Optional is empty in both cases and lead to an error 404;
![firefox_dkhXsPcmhn](https://user-images.githubusercontent.com/42473485/207868570-32329710-21e7-44db-9421-a2718c3d4c81.png)

**Comment by schauder** [comment]
I tried to reproduce the issue based on your information, but failed: https://github.com/schauder/issue-jpa-2720-find-by-uuid

Could you please create a full reproducer? Ideally based on the repository referenced above.

**Comment by burdoto** [comment]
Hello,

As mentioned before [this is the code base state that cause me to publish this issue initially](https://github.com/comroid-git/comroid.org/tree/ab74814b4e71aeecff09a679363c66fa1bd0601f/auth/src/main).
Alternatively, I have [uploaded my reproduction project to GitHub](https://github.com/burdoto/SpringJpaUuidConverter).

Although, I'm starting to think this may be related to me using the MariaDB connector, while the server itself runs on MySQL...

**Comment by burdoto** [comment]
I've tried switching my reproduction project to explicitly MySQL, as the DB Server uses it, but I still get the same results, so I assume this is not the issue.

**Comment by schauder** [comment]
1. The reproducer references a file `/srv/testdb.json` which is not present in the repository.
2. It also seems to expect an existing database. Please use an in memory database or Testcontainers to provide a database for the reproducer.
3. Please include a test reproducing the issue.

**Comment by schauder** [comment]
Since the reproducer works fine, i.e. it does not reproduce the issue, I'll assume this is not a Spring Data issue, but some other issue in your code. 

If you manage to create a reproducer demonstrating an issue with Spring Data JPA, feel free to reopen this issue.
